rubygems/rubygems
# Create Webauthn Listener to retrieve OTP code from browser

_This is being merged into a feature branch, not main, so that work may continue to be developed in small pieces as we built out webauthn for the CLI_

## What was the end-user or developer problem that led to this PR?
Part of implementing Webauthn support for the CLI https://github.com/rubygems/rubygems.org/issues/3453. 

Pivoting from the approach from https://github.com/rubygems/rubygems/pull/6179 as when a user authenticates their security device using a unique link to the browser, the browser will display an OTP code to paste into the CLI. This PR removes the pasting of the code by creating a listener to receive the code.

## What is your fix for the problem, implemented in this PR?



If the user has MFA enabled and has a webauthn device, a socket will be opened on the user's machine. A unique link will be shown to the user with the localhost port of the socket. 

After authentication, the server will redirect back to the socket with the OTP code (eg. `http://localhost:45673?code=Uvh6T57tkWuUnWYo`). The code will be used as the otp option in API requests.

![Screenshot 2023-02-16 at 10 19 26 PM](https://user-images.githubusercontent.com/42748004/219541206-6b52813a-16fd-47be-9539-2f025b301e4e.png)

Note:  if the user has TOTP and Webauthn enabled, the webauthn way of authentication is shown. As a follow up, a fallback option to OTP will be provided.

### Possible responses when receiving the code from the Gem host
1. 200 OK: OTP code was successfully retrieved.
2. 204 No Content: If the request was an OPTIONS request. Since the request was originally sent on the client side of RubyGems.org, a preflight CORS request is sent since this is a cross origin.
3. 400 Bad Request: URL doesn't have code param (eg. `GET http://localhost:45673`)
![Screenshot 2023-02-16 at 10 28 48 PM](https://user-images.githubusercontent.com/42748004/219542339-5330547c-d17c-4ead-bac2-8b2565ea003e.png)

4. 404 Not Found: URL having a bad path (not the root path) (eg. `GET http://localhost:45673/lol?code=Uvh6T57tkWuUnWYo`)
![Screenshot 2023-02-16 at 10 29 42 PM](https://user-images.githubusercontent.com/42748004/219542443-8f240609-e6f0-4d34-82bb-e5e3fbc5da73.png)

5. 405 Method Not Allowed: URL was not a GET/OPTIONS request (eg. `POST http://localhost:45673?code=Uvh6T57tkWuUnWYo`)

## What should reviewers focus on
- If it works with the RubyGems.org changes (TBD)
- Is there an easier way to build socket responses using the std lib?
- Any edge cases missing when receiving a request from the Gem host
- Response bodies both in content type (should we respond in another format like json) and the actual content. Right now we are sending plain text "success" and "missing code parameter" for bad request.

### Testing
1. Rubygems.org code isn't completed yet. You can test this using this branch https://github.com/Shopify/rubygems.org/pull/130. Make sure to setup a user with TOTP and Webauthn.
2. Run `RUBYGEMS_HOST=http://localhost:3000 ruby -Ilib bin/gem signin`
3. Try the same with gem push, yank and owner -a/-r

### Remaining TODOs:

- [x] Add access control headers for 40x responses
- [x] Look into JRuby failing test

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)
